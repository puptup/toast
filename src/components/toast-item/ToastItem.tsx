/* eslint-disable react-hooks/exhaustive-deps */

import { IconsNamesByVariant } from "@constants";
import useAnimation from "@hooks/useAnimation";
import { DispatchEvent } from "@service/eventBus";
import Events from "@service/events";
import { Toast } from "@types";
import React, { useEffect, useState } from "react";

import { CloseButton, IconWrapper, TextWrapper, ToastWrapper } from "./styled";

export interface ToastItemProps {
  toast: Toast;
}

export const remove = (toast: Toast) => {
  DispatchEvent<Toast>(Events.DeleteToast(toast.position), toast);
};

const ToastItem = ({ toast }: ToastItemProps) => {
  const {
    id,
    type,
    timeToDelete,
    title,
    spaceToEdge,
    spaceBetweenToasts,
    description,
    animationIn,
    animationOut,
    animationDuration,
  } = toast;

  const { keyframeIn, keyframeOut } = useAnimation(animationIn, animationOut);

  const [out, setOut] = useState(false);

  const handleOutAnimationAndRemove = () => {
    setOut(true);
    setTimeout(() => {
      remove(toast);
    }, animationDuration / 2);
  };

  useEffect(() => {
    setTimeout(() => {
      handleOutAnimationAndRemove();
    }, timeToDelete + animationDuration);
  }, []);

  const iconName = IconsNamesByVariant[type];

  return (
    <div id={id}>
      <ToastWrapper
        animationIn={keyframeIn}
        animationOut={keyframeOut}
        animationDuration={animationDuration}
        out={out}
        variant={type}
        spaceToEdge={spaceToEdge}
        spaceBetweenToasts={spaceBetweenToasts}
        data-testid={`toast-${id}`}
      >
        <IconWrapper className="material-icons" variant={type}>
          {iconName}
        </IconWrapper>
        <TextWrapper>
          <p>{title}</p>
          <p>{description}</p>
        </TextWrapper>
        <CloseButton type="button" onClick={handleOutAnimationAndRemove}>
          <IconWrapper className="material-icons" variant={type}>
            close
          </IconWrapper>
        </CloseButton>
      </ToastWrapper>
    </div>
  );
};

export default React.memo(ToastItem);
