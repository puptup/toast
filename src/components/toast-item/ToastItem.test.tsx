import "@testing-library/jest-dom";

import { defaultToast } from "@constants";
import { render } from "@testing-library/react";
import { Toast } from "@types";
import React from "react";

import ToastItem from "./ToastItem";

const infoToast: Toast = {
  ...defaultToast,
  id: "test",
  type: "info",
};

const warningToast: Toast = {
  ...defaultToast,
  id: "test",
  type: "info",
};

const successToast: Toast = {
  ...defaultToast,
  id: "test",
  type: "success",
};

const errorToast: Toast = {
  ...defaultToast,
  id: "test",
  type: "error",
};

describe("Toast", () => {
  it("should be info toast", () => {
    const { queryByTestId } = render(<ToastItem toast={infoToast} />);
    const element = queryByTestId("toast-test");
    expect(element).toMatchSnapshot();
  });

  it("should be warning toast", () => {
    const { queryByTestId } = render(<ToastItem toast={warningToast} />);
    const element = queryByTestId("toast-test");
    expect(element).toMatchSnapshot();
  });

  it("should be success toast", () => {
    const { queryByTestId } = render(<ToastItem toast={successToast} />);
    const element = queryByTestId("toast-test");
    expect(element).toMatchSnapshot();
  });

  it("should be error toast", () => {
    const { queryByTestId } = render(<ToastItem toast={errorToast} />);
    const element = queryByTestId("toast-test");
    expect(element).toMatchSnapshot();
  });
});
