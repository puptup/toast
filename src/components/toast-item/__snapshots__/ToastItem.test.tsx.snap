// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast should be error toast 1`] = `
<div
  class="sc-bcXHqe rQoTR"
  data-testid="toast-test"
>
  <div
    class="sc-hLBbgP hClGEU material-icons"
  >
    bug_report
  </div>
  <div
    class="sc-gswNZR etjVmh"
  >
    <p>
      Toast
    </p>
    <p>
      Toast description
    </p>
  </div>
  <button
    class="sc-dkrFOg Rsggx"
    type="button"
  >
    <div
      class="sc-hLBbgP hClGEU material-icons"
    >
      close
    </div>
  </button>
</div>
`;

exports[`Toast should be info toast 1`] = `
<div
  class="sc-bcXHqe fQnZmo"
  data-testid="toast-test"
>
  <div
    class="sc-hLBbgP hClGEU material-icons"
  >
    info
  </div>
  <div
    class="sc-gswNZR etjVmh"
  >
    <p>
      Toast
    </p>
    <p>
      Toast description
    </p>
  </div>
  <button
    class="sc-dkrFOg Rsggx"
    type="button"
  >
    <div
      class="sc-hLBbgP hClGEU material-icons"
    >
      close
    </div>
  </button>
</div>
`;

exports[`Toast should be success toast 1`] = `
<div
  class="sc-bcXHqe ckrIQx"
  data-testid="toast-test"
>
  <div
    class="sc-hLBbgP hClGEU material-icons"
  >
    done
  </div>
  <div
    class="sc-gswNZR etjVmh"
  >
    <p>
      Toast
    </p>
    <p>
      Toast description
    </p>
  </div>
  <button
    class="sc-dkrFOg Rsggx"
    type="button"
  >
    <div
      class="sc-hLBbgP hClGEU material-icons"
    >
      close
    </div>
  </button>
</div>
`;

exports[`Toast should be warning toast 1`] = `
<div
  class="sc-bcXHqe fQnZmo"
  data-testid="toast-test"
>
  <div
    class="sc-hLBbgP hClGEU material-icons"
  >
    info
  </div>
  <div
    class="sc-gswNZR etjVmh"
  >
    <p>
      Toast
    </p>
    <p>
      Toast description
    </p>
  </div>
  <button
    class="sc-dkrFOg Rsggx"
    type="button"
  >
    <div
      class="sc-hLBbgP hClGEU material-icons"
    >
      close
    </div>
  </button>
</div>
`;
